packages/fiber/src/core/reconciler.tsx

```diff
@@ -337,9 +337,14 @@ function handleContainerEffects(parent: Instance, child: Instance, beforeChild?:
 
 function appendChild(parent: HostConfig['instance'], child: HostConfig['instance'] | HostConfig['textInstance']) {
   if (!child) return
-
   // Link instances
   child.parent = parent
+
+  // Remove existing child if it exists
+  const existingIndex = parent.children.indexOf(child)
+  if (existingIndex !== -1) parent.children.splice(existingIndex, 1)
+
+  // Append child
   parent.children.push(child)
```

Comment:
https://github.com/pmndrs/react-three-fiber/pull/3490#issuecomment-2766740010 Maybe we can move the check before overriding the parent.

```ts
// Remove existing child if it exists
if (child.parent === parent) {
  const existingIndex = parent.children.indexOf(child)
  if (existingIndex !== -1) parent.children.splice(existingIndex, 1)
}

// Link instances
child.parent = parent

// Append child
parent.children.push(child)
```

Reply:
Yes this makes sense. I think further optimizations would require convincing React to provide a path for append separate from reorder since it has this information internally. However, for now we can probably not worry about this.